### PR TITLE
Jm simple ms teams

### PR DESCRIPTION
**Pull request by JakeMathews**
IALERT-2163
Adds support for MS Teams with the new channel implementation.

I created a test that is ummm pretty bad. Which is why it's not in this PR:
https://github.com/blackducksoftware/blackduck-alert/compare/jm-simple-ms-teams...jm-ms-teams-test
A new end-to-end test framework for these channels is planned with IALERT-1973, but this test was my curious attempt to see some results.
image

### CODE REVIEW

**Review by jamesrichard91 — APPROVED** [review]

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public MSTeamsChannelMessageSender(RestChannelUtility restChannelUtility, MsTeamsKey msTeamsKey) {
+        this.restChannelUtility = restChannelUtility;
+        this.msTeamsKey = msTeamsKey;
+    }
+
+    @Override
+    public MessageResult sendMessages(MSTeamsJobDetailsModel msTeamsJobDetailsModel, List<MSTeamsChannelMessageModel> channelMessages) throws AlertException {
+        String webhook = msTeamsJobDetailsModel.getWebhook();
+
+        String messageTitle = "Received a message from Black Duck"; // TODO: How should we get the provider? Each message converted in a ChannelMessageConverter is as ProviderMessage.
```
Comment (JakeMathews) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:57]: Would grouping the messages by provider mess up the order? What if messages were grouped by provider into a request in the order they are in. 
```
messagesByProvider = [ BlackDuck, BlackDuck, TestProvider, BlackDuck ]
requestsByProvider = [ BlackDuck, TestProvider, BlackDuck ]
```

Comment (gkillough) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:57]: A `ProviderMessageHolder` will only ever have messages from the same provider (and more specifically, provider-config). We could add the provider as a first-class citizen of that object to express that more clearly.

**Review by ChomickiM — APPROVED** [review]

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageConverter.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return createMessageModel(messageChunks);
+    }
+
+    @Override
+    protected List<MSTeamsChannelMessageModel> convertProjectMessageToChannelMessages(MSTeamsJobDetailsModel msTeamsJobDetailsModel, ProjectMessage projectMessage, List<String> messageChunks) {
+        return createMessageModel(messageChunks);
+    }
+
+    private List<MSTeamsChannelMessageModel> createMessageModel(List<String> messageChunks) {
+        // The title for each message is currently empty because the old implementation would attempt to build a title based on components, or default to an empty String.
```
Comment (bamandel) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageConverter.java:56]: You should have access to component data in the projectMessage where this component data lives, do we not want to try to mimic the functionality using that?

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        String jsonString = getJsonString(messageTitle, messages);
+        return restChannelUtility.createPostMessageRequest(webhook, requestHeaders, jsonString);
+    }
+
+    private String getJsonString(String title, List<MSTeamsChannelMessageModel> messageSections) {
+        JsonObject json = new JsonObject();
+        json.addProperty("@type", "MessageCard");
+        json.addProperty("@context", "https://schema.org/extensions");
+        json.addProperty("summary", "New Content from Alert");
+        json.addProperty("themeColor", "5A2A82");
```
Comment (bamandel) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:78]: I think at least the color hex code should be in a static variable with a name for the color.

**Review by bamandel — APPROVED** [review]

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java (hunk 3 above)
Comment (JakeMathews) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java]: How's this?
```
    private static final String MESSAGE_THEME_COLOR = "5A2A82"; // Synopsys Purple
    private static final String MESSAGE_SUMMARY = "New Content from Alert";
 ```

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageFormatter.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return String.format("**%s**", txt);
+    }
+
+    @Override
+    public String createLink(String txt, String url) {
+        return String.format("[%s](%s)", txt, url);
+    }
+
+    @Override
+    public String getLineSeparator() {
```
Comment (gkillough) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageFormatter.java:70]: This looks redundant. We might have to make the getter on the superclass final.

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java (hunk 3 above)
Comment (bamandel) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java]: I think that's good. Adding the comment as part of the name may be better as you'll see the color directy where it's used.

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageConverter.java (hunk 2 above)
Comment (JakeMathews) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageConverter.java:56]: You're right. Upon looking at the current implementation, the logic around titles is quite complex. I will try to capture the same functionality.

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageFormatter.java (hunk 4 above)
Comment (JakeMathews) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageFormatter.java]: I made all the getters in ChannelMessageFormatter `final`

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java (hunk 1 above)
Comment (gkillough) [reply channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:60]: I created a separate issue in Jira for this.

**Review by gkillough — APPROVED** [review]

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+        Map<String, String> requestHeaders = new HashMap<>();
+        requestHeaders.put("Content-Type", "application/json");
+
+        String provider = channelMessages.stream()
+                              .findFirst()
+                              .map(MSTeamsChannelMessageModel::getProviderDetails)
+                              .map(ProviderDetails::getProvider)
+                              .map(LinkableItem::getValue)
+                              .orElse("Black Duck"); // TODO: Should we always assume we have at least one channel message to avoid this Optional?
```
Comment (gkillough) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:70]: I would put it in a private method for readability, but we can remove the TODO. I think it's fine to provide a default. Also, I think we are supposed to have one word for "BlackDuck".

File: channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+        return new MessageResult(String.format("Successfully sent %d MSTeams message(s)", channelMessages.size()));
+    }
+
+    private Request createRequestsForMessage(String webhook, String messageTitle, List<MSTeamsChannelMessageModel> messages, Map<String, String> requestHeaders) {
+        String jsonString = getJsonString(messageTitle, messages);
+        return restChannelUtility.createPostMessageRequest(webhook, requestHeaders, jsonString);
+    }
+
+    private String getJsonString(String title, List<MSTeamsChannelMessageModel> messageSections) {
```
Comment (gkillough) [inline channel/src/main/java/com/synopsys/integration/alert/channel/msteams2/MSTeamsChannelMessageSender.java:84]:
```suggestion
        String jsonString = createJsonString(messageTitle, messages);
        return restChannelUtility.createPostMessageRequest(webhook, requestHeaders, jsonString);
    }

    private String createJsonString(String title, List<MSTeamsChannelMessageModel> messageSections) {
```

**Pull request merged**